plugins/experimental/statichit/statichit.cc

```diff
@@ -0,0 +1,637 @@
+      HeaderFieldDateSet(response, TS_MIME_FIELD_LAST_MODIFIED, TS_MIME_LEN_LAST_MODIFIED, time(nullptr));
+
+    } else {
+      HeaderFieldStringSet(response, TS_MIME_FIELD_CACHE_CONTROL, TS_MIME_LEN_CACHE_CONTROL, "no-cache");
+    }
+
+    HeaderFieldStringSet(response, TS_MIME_FIELD_CONTENT_TYPE, TS_MIME_LEN_CONTENT_TYPE, trq->mimeType.c_str());
+  }
+
+  // Write the header to the IO buffer. Set the VIO bytes so that we can get a WRITE_COMPLETE
+  // event when this is done.
+  int hdrlen = TSHttpHdrLengthGet(response.buffer, response.header);
+
+  TSHttpHdrPrint(response.buffer, response.header, trq->writeio.iobuf);
+  TSVIONBytesSet(trq->writeio.vio, hdrlen);
+  TSVIOReenable(trq->writeio.vio);
+
+  TSStatIntIncrement(StatCountBytes, hdrlen);
+}
+
+static bool
+StaticHitParseRequest(StaticHitRequest *trq)
+{
+  const char *path;
+  int pathsz;
+
+  // Make sure this is a GET request
+  path = TSHttpHdrMethodGet(trq->rqheader.buffer, trq->rqheader.header, &pathsz);
+  if (path != TS_HTTP_METHOD_GET) {
+    VDEBUG("%.*s method is not supported", pathsz, path);
+    return false;
+  }
+
+  return true;
+}
+
+// Handle events from TSHttpTxnServerIntercept. The intercept
+// starts with TS_EVENT_NET_ACCEPT, and then continues with
+// TSVConn events.
+static int
+StaticHitInterceptHook(TSCont contp, TSEvent event, void *edata)
+{
+  VDEBUG("StaticHitInterceptHook: %p ", edata);
+
+  argument_type arg(edata);
+
+  VDEBUG("contp=%p, event=%s (%d), edata=%p", contp, TSHttpEventNameLookup(event), event, arg.ptr);
+
+  switch (event) {
+  case TS_EVENT_NET_ACCEPT: {
+    // TS_EVENT_NET_ACCEPT will be delivered when the server intercept
+    // is set up by the core. We just need to allocate a statichit
+    // request state and start reading the VC.
+    StaticHitRequest *trq = static_cast<StaticHitRequest *>(TSContDataGet(contp));
+
+    TSStatIntIncrement(StatCountResponses, 1);
+    VDEBUG("allocated server intercept statichit trq=%p", trq);
+
+    // This continuation was allocated in StaticHitTxnHook. Reset the
+    // data to keep track of this generator request.
+    TSContDataSet(contp, trq);
+
+    // Start reading the request from the server intercept VC.
+    trq->readio.read(arg.vc, contp);
+    VIODEBUG(trq->readio.vio, "started reading statichit request");
+
+    return TS_EVENT_NONE;
+  }
+
+  case TS_EVENT_NET_ACCEPT_FAILED: {
+    // TS_EVENT_NET_ACCEPT_FAILED will be delivered if the
+    // transaction is cancelled before we start tunnelling
+    // through the server intercept. One way that this can happen
+    // is if the intercept is attached early, and then we serve
+    // the document out of cache.
+
+    // There's nothing to do here except nuke the continuation
+    // that was allocated in StaticHitTxnHook().
+
+    StaticHitRequest *trq = static_cast<StaticHitRequest *>(TSContDataGet(contp));
+    delete trq;
+
+    TSContDestroy(contp);
```

Comment:
Should this be done via `StaticHitRequestDestroy`? It was hard to track down where the `cont` is destroyed in the non-error case.